include/cute/arch/copy_xe.hpp

```diff
@@ -1,45 +1,201 @@
+    int pitch_minus_one, int2_ coord, uint8 data));
+SYCL_DEVICE_BUILTIN(ushort8 __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+SYCL_DEVICE_BUILTIN(uint8 __builtin_IB_subgroup_block_read_flat_u32_m8k16v1(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+
+/// Load A
+SYCL_DEVICE_BUILTIN(ushort64 __builtin_IB_subgroup_block_read_flat_u16_m32k16v2(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+SYCL_DEVICE_BUILTIN(ushort32 __builtin_IB_subgroup_block_read_flat_u16_m16k16v2(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+SYCL_DEVICE_BUILTIN(ushort16 intel_subgroup_block_read_u16_m8k16v2(
+    __global void *baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+SYCL_DEVICE_BUILTIN(ushort32 __builtin_IB_subgroup_block_read_flat_u16_m32k16v1(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+
+/// Load B
+SYCL_DEVICE_BUILTIN(uint16 __builtin_IB_subgroup_block_read_flat_u32_m16k16v1(
+    long baseoffset, int width_minus_one, int height_minus_one,
+    int pitch_minus_one, int2_ coord));
+
 #undef SYCL_DEVICE_BUILTIN
 
-struct XE_2D_LOAD //m8k16
-{
-    template<class T>
-    CUTE_HOST_DEVICE static void copy(const void* baseoffset, int width, int height, int pitch, int2_ coord, T* dst)
-    {
-        if constexpr(sizeof(T)==sizeof(ushort)) {
-            *(ushort8*)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1((long)baseoffset, width - 1, height - 1, pitch - 1, coord);    
-        } else if constexpr(sizeof(T)==sizeof(uint)) {
-            *(uint8*)dst = __builtin_IB_subgroup_block_read_flat_u32_m8k16v1((long)baseoffset, width - 1, height - 1, pitch - 1, coord);
-        } else {
-            static_assert(false);
-        }
-    }
-};
-
-struct XE_2D_SAVE //m8k16
-{
-    template<class T>
-    CUTE_HOST_DEVICE static void copy(void* baseoffset, int width, int height, int pitch, int2_ coord, const T* src)
-    {
-        if constexpr(sizeof(T)==sizeof(uint)) {
-            __builtin_IB_subgroup_block_write_flat_u32_m8k16v1((long)baseoffset, width - 1, height - 1, pitch - 1, coord, *(uint8*)src);
-        } else {
-            static_assert(false);
-        }
-    }
+struct XE_2D_U16X8X16X1X1_LD_N
+{
+  template <class T>
+  CUTE_HOST_DEVICE static void copy(const void *baseoffset, int width,
+                                    int height, int pitch, int2_ coord,
+                                    T *dst) {
+    static_assert(sizeof(T) == 2, "Expected T to have size 2");
+    *(ushort8 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
+        (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
```

Comment:
Should we have error checking for invalid pipeline creation, similar to how Nvidia has done for their various architectures, i.e.
```
#if defined(ARCH_PVC)
*(ushort8 *)dst = __builtin_IB_subgroup_block_read_flat_u16_m8k16v1(
        (long)baseoffset, width - 1, height - 1, pitch - 1, coord);
#else
CUTE_INVALID_CONTROL_PATH("Trying to use block loads on non-PVC hardware")
#endif
```

Reply:
Similarly for all other structs as well